autoload/lsp/internal/semantic.vim

```diff
@@ -0,0 +1,216 @@
+let s:use_vim_textprops = lsp#utils#_has_textprops() && !has('nvim')
+let s:use_nvim_highlight = exists('*nvim_buf_add_highlight') && has('nvim')
+let s:textprop_cache = 'vim-lsp-semantic-cache'
+
+if s:use_nvim_highlight
+    let s:namespace_id = nvim_create_namespace('vim-lsp-semantic')
+endif
+
+if !hlexists('LspUnknownScope')
+    highlight LspUnknownScope gui=NONE cterm=NONE guifg=NONE ctermfg=NONE guibg=NONE ctermbg=NONE
+endif
+
+" Global functions {{{1
+function! lsp#internal#semantic#is_enabled() abort
+    return g:lsp_semantic_enabled && (s:use_vim_textprops || s:use_nvim_highlight) ? v:true : v:false
+endfunction
+
+function! lsp#internal#semantic#_enable() abort
+    if !lsp#internal#semantic#is_enabled() | return | endif
+
+    " request and process full semantic tokens refresh when the filetype changes
+    " or when the text is modified
+    let s:Dispose = lsp#callbag#pipe(
+        \     lsp#callbag#fromEvent(['FileType', 'TextChanged', 'TextChangedI']),
+        \     lsp#callbag#filter({_->lsp#internal#semantic#is_enabled()}),
+        \     lsp#callbag#debounceTime(g:lsp_semantic_delay),
+        \     lsp#callbag#filter({_->getbufvar(bufnr('%'), '&buftype') !~# '^(help\|terminal\|prompt\|popup)$'}),
```

Comment:
might not be required now, but in the future we may have to add max buffer size flag so it resets highlights but skips making request to semantic highlights for perf. Some languages such as c/c++ has large number of lines.

Reply:
https://github.com/prabirshrestha/asyncomplete-buffer.vim/blob/018bcf0f712ce0fde3f1f2eaabd7004fccb2d34a/autoload/asyncomplete/sources/buffer.vim#L34-L48

Reply:
Done in [7a4c5b1](https://github.com/prabirshrestha/vim-lsp/pull/1275/commits/7a4c5b154835d6aee22dc859cc5c13495ca64c5e). 

Semantic highlighting could be implemented using the `textDocument/semanticTokens/range` request for only the visible region on large files, however server support for this is patchy.